packages/@lwc/ssr-runtime/src/render.ts

```diff
@@ -39,22 +58,80 @@ export function* fallbackTmpl(
     }
 }
 
+export function fallbackTmplNoYield(
+    emit: (segment: string) => void,
+    _props: unknown,
+    _attrs: unknown,
+    _slotted: unknown,
+    Cmp: LightningElementConstructor,
+    _instance: unknown
+) {
+    if (Cmp.renderMode !== 'light') {
+        emit('<template shadowrootmode="open"></template>');
+    }
+}
+
 export type GenerateMarkupFn = (
     tagName: string,
     props: Record<string, any> | null,
     attrs: Attributes | null,
     slotted: Record<number | string, AsyncGenerator<string>> | null
 ) => AsyncGenerator<string>;
 
+export type GenerateMarkupFnAsyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => Promise<void>;
+
+export type GenerateMarkupFnSyncNoGen = (
+    emit: (segment: string) => void,
+    tagName: string,
+    props: Record<string, any> | null,
+    attrs: Attributes | null,
+    slotted: Record<number | string, AsyncGenerator<string>> | null
+) => void;
+
+type GenerateMarkupFnVariants =
+    | GenerateMarkupFn
+    | GenerateMarkupFnAsyncNoGen
+    | GenerateMarkupFnSyncNoGen;
+
 export async function serverSideRenderComponent(
     tagName: string,
-    compiledGenerateMarkup: GenerateMarkupFn,
-    props: Record<string, any>
+    compiledGenerateMarkup: GenerateMarkupFnVariants,
+    props: Record<string, any>,
+    mode: 'asyncYield' | 'async' | 'sync' = 'asyncYield'
 ): Promise<string> {
     let markup = '';
 
-    for await (const segment of compiledGenerateMarkup(tagName, props, null, null)) {
-        markup += segment;
+    if (mode === 'asyncYield') {
+        for await (const segment of (compiledGenerateMarkup as GenerateMarkupFn)(
```

Comment:
Oh that's an interesting idea. Your idea here wouldn't address the (additional) footgun where different components were compiled in different modes. I guess I didn't imagine these different modes surviving forever, and we'd just pick one in a couple of months once everybody agreed on the path forward. But I kind of like the idea of detecting this and avoiding the footgun (and type casting).

If you don't mind, I'd like to add that in a follow-up PR, since it'll be a somewhat invasive change.

Reply:
If we pick sync-no-yield then we're going to change the signature/return type of the entire `serverSideRenderComponent` function anyway. I wouldn't sweat this for now, but it would make our life easier if we don't pick something soonish.